src/shm/pool/multi.rs

```diff
@@ -0,0 +1,203 @@
+        self.inner.resize(size, conn)
+    }
+    fn free(&self, buffer: &wl_buffer::WlBuffer) {
+        if let Some(buffer) = self.buffer_list.iter().find(|b| b.buffer.eq(buffer)) {
+            buffer.free.swap(true, Ordering::Relaxed);
+        }
+    }
+    /// Removes the buffer with the given key from the pool and rearranges the others
+    pub fn remove(&mut self, key: &K) {
+        if let Some((i, buffer)) =
+            self.buffer_list
+            .iter()
+            .enumerate()
+            .find(|b| b.1.key.eq(key)) {
+            let mut offset = buffer.offset;
+            self.buffer_list.remove(i);
+            for buffer in &mut self.buffer_list {
+                if buffer.offset > offset && buffer.free.load(Ordering::Relaxed) {
+                    let l_offset = buffer.offset;
+                    buffer.offset = offset;
+                    offset = l_offset;
+                } else {
+                    break;
+                }
+            }
+        }
+    }
+    /// Returns the buffer associated with the given key.
+    /// If it's not possible to use the buffer associated with the key, None is returned.
+    pub fn create_buffer<D, U>(
+        &mut self,
+        width: i32,
+        stride: i32,
+        height: i32,
+        key: &K,
+        format: wl_shm::Format,
+        udata: U,
+        conn: &mut ConnectionHandle,
+        qh: &QueueHandle<D>,
+    ) -> Option<(usize, wl_buffer::WlBuffer, &mut [u8])>
+    where
+        D: Dispatch<wl_buffer::WlBuffer, UserData = U> + 'static,
+        U: Send + Sync + 'static,
+    {
+        let mut found_key = false;
+        let mut offset = 0;
+        let size = (stride * height) as usize;
+        let mut index = None;
+
+        // This loop serves to found the buffer associated to the key.
+        for (i, buffer) in self.buffer_list.iter_mut().enumerate() {
+            if buffer.key.eq(key) {
+                found_key = true;
+                if buffer.free.load(Ordering::Relaxed) {
+                    // Increases the size of the Buffer if it's too small and add 5% padding
+                    buffer.size = buffer.size.max(size + size / 20);
+                    index = Some(i);
+                }
+            // If a buffer is resized, it is likely that the followings might overlap
+            } else if offset > buffer.offset {
+                // When the buffer is free, it's safe to shift it because we know the compositor won't try to read it.
+                if buffer.free.load(Ordering::Relaxed) {
+                    buffer.offset = offset;
+                } else {
+                    // If one of the overlapping buffers is busy, then no buffer can be returned because it could result in a data race.
+                    index = None;
+                }
+            } else if found_key {
+                break;
+            }
+            offset += buffer.size;
+        }
```

Comment:
> It seems to me that this is actually fine, because the method will just come back and fix its state on the next invocation, meaning that when it returns Some(...), the state is actually correct. Is that indeed your intent?

Yes. I'll comment that. Also I found [this](http://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/first-edition/documentation.html) for documentation and it seems to cover everything I need but is there a way to preview the actual web page without compulsively running `cargo doc`?

> Then at a later point it redraws B, and then later again it tries to redraw its main surface, and A is still unusable because B is again in use. Following that pattern we can imagine unlucky cases where some buffers in the pool will never be usable again and end up as dead space.

I think if the client puts a cap of how much buffers they want, they can use frame callbacks to schedule another draw and by then I really doubt the buffer will still be in use. Worse case scenario, if the client doesn't have a hard cap, it should at least know when there's too much buffers for a surface and remove them manually.

Reply:
> is there a way to preview the actual web page without compulsively running `cargo doc`?

I don't know about that sorry, I just run `cargo doc --no-deps --open` when I need to check how my documentation looks.

Reply:
cargo-watch appears to exist for live reload, but I haven't used it before.

Reply:
I'll give it a shot, thanks.